site/cds_rdm/ldap/utils.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# CDS-RDM is free software; you can redistribute it and/or modify it under
+# the terms of the MIT License; see LICENSE file for more details.
+
+"""CDS-RDM ldap serializers."""
+
+from functools import partial
+
+from cds_rdm.ldap.errors import InvalidLdapUser
+from invenio_accounts.models import User
+from invenio_oauthclient.models import UserIdentity
+from invenio_userprofiles import UserProfile
+
+
+def serialize_ldap_user(ldap_user_data, log_func=None):
+    """Create ldap user."""
+
+    def serialize(ldap_user_data):
+        decoded_data = {}
+        for key, value in ldap_user_data.items():
+            # get first value of each field from LDAP
+            # decode bytes to UTF-8
+            decoded_data[key] = value[0].decode("utf8")
+        serialized_data = dict(
+            user_email=decoded_data["mail"].lower(),
+            user_profile_full_name=decoded_data["displayName"],
+            user_identity_id=decoded_data["uidNumber"],
+            cern_account_type=decoded_data["cernAccountType"],
+            remote_account_person_id=str(decoded_data["employeeID"]),
+            remote_account_department=decoded_data["department"],
+        )
+
+        return serialized_data
+
+    def validate_required(ldap_user_data, employee_id, log_func):
+        if "mail" not in ldap_user_data:
+            log_func_missing_email = partial(
+                log_func, extra=dict(employee_id=employee_id)
+            )
+            raise InvalidLdapUser(
+                f"LDAP user with employeeID {employee_id}" f"has no email address.",
+                log_func=log_func_missing_email,
+            )
+
+    def validate_user_data(serialized_ldap_user_data, employee_id, log_func):
+        """Validate user data values."""
+        email = serialized_ldap_user_data["user_email"]
+        # check if email is empty string.
+        # It happens when the account is not fully created on LDAP
+        if not email:
+            log_func_missing_email = partial(
+                log_func, extra=dict(employee_id=employee_id)
+            )
+            raise InvalidLdapUser(
+                f"LDAP user with employeeID {employee_id}" f" has no email address.",
+                log_func=log_func_missing_email,
+            )
+
+    employee_id = ldap_user_data["employeeID"][0].decode("utf8")
+    try:
+        validate_required(ldap_user_data, employee_id, log_func)
+        serialized_data = serialize(ldap_user_data)
+        validate_user_data(serialized_data, employee_id, log_func)
+        return serialized_data
+    except InvalidLdapUser:
+        return
+
+
+def user_exists(ldap_user):
+    """Check if user exists in the db."""
+    if not ldap_user:
+        return False
+
+    email = ldap_user["user_email"]
+    user_email_exists = User.query.filter_by(email=email).count() > 0
```

Comment:
I _think_ a `count` query is more expensive than a `one_or_none` since it will do a full pass on the db table. `email` has a unique constraint in invenio accounts so there could not be more than one.

To be checked, does sqlalchemy ORM have an `exists()`, it might be more optimized for this use case.